jax/_src/scipy/signal.py

```diff
@@ -150,21 +150,30 @@ def _fftconvolve_unbatched(in1: Array, in2: Array, mode: str) -> Array:
 # Note: we do not re-use the code from jax.numpy.convolve here, because the handling
 # of padding differs slightly between the two implementations (particularly for
 # mode='same').
-def _convolve_nd(in1: Array, in2: Array, mode: str, *, precision: PrecisionLike) -> Array:
+def _convolve_nd(in1: Array, in2: Array, mode: str, boundary: str = 'fill', *,
+                 precision: PrecisionLike) -> Array:
   if mode not in ["full", "same", "valid"]:
     raise ValueError("mode must be one of ['full', 'same', 'valid']")
   if in1.ndim != in2.ndim:
     raise ValueError("in1 and in2 must have the same number of dimensions")
   if in1.size == 0 or in2.size == 0:
     raise ValueError(f"zero-size arrays not supported in convolutions, got shapes {in1.shape} and {in2.shape}.")
+  if boundary == 'wrap' and in1.ndim != 2:
+      raise NotImplementedError("boundary='wrap' is implemented only for 2D inputs")
   in1, in2 = promote_dtypes_inexact(in1, in2)
 
   no_swap = all(s1 >= s2 for s1, s2 in zip(in1.shape, in2.shape))
   swap = all(s1 <= s2 for s1, s2 in zip(in1.shape, in2.shape))
-  if not (no_swap or swap):
-    raise ValueError("One input must be smaller than the other in every dimension.")
+  if mode == 'valid' and not (no_swap or swap):
+    raise ValueError("For 'valid' mode, one input must always be smaller than the other in every dimension.")
 
+  shape1 = in1.shape
   shape_o = in2.shape
+  if in1.ndim > 1:
+    pad_width = tuple(map(max, shape1, shape_o))
+    if boundary == 'wrap' and mode != 'valid':
+      in1 = jnp.pad(in1, pad_width, mode='wrap')
```

Comment:
This is doing unnecessary work when `boundary != 'wrap'`.

Also it seems like it would simplify things to just generalize this to `N` dimensions (even if only the 2D path is used). What if we did something like this here:
```python
out_slices = ()
if boundary == 'wrap' and mode != 'valid':
  pad_width = tuple(map(max, shape1, shape_o))
  in1 = jnp.pad(in1, pad_width, mode='wrap')
  if mode == 'same':
    slices = tuple(slice(p:p+s) for p, s in zip(pad_width, shape1)
  elif mode == 'full':
    slices = tuple(slice(p:p+s1+s2) for p, s1, s2 in zip(pad_width, shape1, shape_o)
```
Then below, instead of multiple return statements we could just write:
```python
return result[0, 0, *slices]
```
The nice thing about this is that it removes the need for "action-at-a-distance", which can lead to harder-to-maintain code. For example, in the current version, `pad_width` is conditionally defined, and then below is conditionally referenced. This is a bit of a code smell, because it means we need to keep the two conditions in sync. My proposal removes that action at a distance.

What do you think?

Reply:
Thank you for the suggestion. It appears to be a good idea to generalize for `N` dimensions. I have tested it with the existing test cases and it passes all the tests locally.

In that case, we can avoid raising the `NotImplementedError` for 1D inputs, since we are generalizing it.